### PR TITLE
Feature/issue 848 map rect mpi, fixes #848

### LINKED ISSUE #848: map_rect MPI
#### Summary:
Introduce MPI enabled map_rect backend

#### Description:
enable within-chain parallelization in stan using MPI in map_rect.

The map_rect facility allows to specify N independent function evaluations at once and this issue is meant to establish a parallel execution mechanism through the use of MPI.

#### Reproducible Steps:
NA

#### Current Output:
all good, bbut too slow for some applications like ODE based hierarchical models

#### Expected Output:
faster output with more CPUs

#### Additional Information:
Provide any additional information here.

#### Current Version:
v2.17.0

### PR DESCRIPTION
**Pull request by wds15**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
Add MPI backend to map_rect which executes tasks in parallel.

#### Intended Effect:
Use multiple cores to execute independent tasks passed to map_rect.

#### How to Verify:
Please first enable MPI, see the stan-math wiki for doing so. The tests to run are

- `test/unit/math/{prim,rev}/mat/functor/map_rect_mpi_test.cpp`
- `test/unit/math/prim/mat/functor/mpi_parallel_call_test.cpp`

#### Side Effects:
Introduces the new `mpi_parallel_call` facility which comes with its own doc and tests.

#### Documentation:
Lot's of doxygen documentation.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by wds15** [comment]
Cool. Now it's ready to go. I had to catch a few boundary cases for the code for which I added tests.

This is not the easiest thing to review, I am afraid. I tried to doc things quite a bit using doxygen and I hope this helps. Yet, keeping track of what happens where when and how caching and exceptions are handled is not trivial. So if we need a call over this, then we can arrange that.

**Comment by yizhang-cae** [comment]
https://github.com/stan-dev/math/blob/8c87ffc3fd84bb0e91cb2af43a4b45888be4183e/stan/math/prim/mat/functor/mpi_parallel_call.hpp#L295
Is it necessary for all processes to gather? Workers just need to be informed by the root to return or not.

**Comment by yizhang-cae** [comment]
All you need is `all_ok`. Let root do a reduce, and broadcast `all_ok`.

Regards,
Yi Zhang, Ph.D.
Metrum Research Group LLC
2 Tunxis Road, Tariffville, CT 06081
yiz@metrumrg.com
www.metrumrg.com

**Comment by wds15** [comment]
Yeah, I can do that:

1. gather status at root
2. broadcast boolean all_ok flag

Fine with me.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        chunks_result[i] += world_f_out[ij] * num_outputs_per_job_;
+
+    // collect results on root
+    boost::mpi::gatherv(world_, local_output.data(), chunks_result[rank_],
+                        world_result.data(), chunks_result, 0);
+
+    // let root know if all went fine everywhere
+    int cluster_status = 0;
+    boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+    bool all_ok = cluster_status == static_cast<int>(world_size_);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I think it `local_ok` being bad will cause `all_ok` to be bad?  If so, I don't know why it's being called here.  If this ust sets `local_ok`, then it doesn't get used anywhere and probably should.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+  void scatter_array_2d_cached(typename T_cache::cache_t& data) {
+    // distribute data only if not in cache yet
+    if (T_cache::is_valid()) {
+      return;
+    }
+
+    // number of elements of each data item must be transferred to
+    // the workers
+    std::vector<int> data_dims = dims(data);
+    data_dims.resize(2);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I think this can just be

```
std::vector<int> data_dims(data.begin(), data.begin() + 2);
```
I think you could also just bypass the vector storage by declaring
```
int* data_dims = { data[0], data[1] };
```
But I'd really be inclined just to name the entries because they get pulled out again as `data_dims[0]` and `data_dims[1]` later and by naming them, you auto-doc them.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+      local_data.push_back(data_elem);
+      local_iter += data_dims[1];
+    }
+
+    // finally we cache it locally
+    T_cache::store(local_data);
+  }
+
+  template <typename T_cache>
+  void broadcast_1d_cached(typename T_cache::cache_t& data) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/functor/mpi_parallel_call.hpp]: needs doc.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    if (T_cache::is_valid()) {
+      return;
+    }
+
+    std::size_t data_size = data.size();
+    boost::mpi::broadcast(world_, data_size, 0);
+
+    typename std::remove_cv<typename T_cache::cache_t>::type local_data = data;
```
Comment (bob-carpenter) [inline stan/math/prim/mat/functor/mpi_parallel_call.hpp]: This is a little too fancy for me with the removal of qualifiers.  Can you just declare this locally?

Also, rather than assigning to `data` then resizing, it'd be better just to use iterators to set it to the right things to begin with if it's a `std::vector` under all that typing.

File: test/unit/math/rev/mat/functor/map_rect_mpi_test.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+      z_var2.insert(z_var2.end(), shared_params_v2_vec.begin(),
+                    shared_params_v2_vec.end());
+      z_var2.insert(z_var2.end(), job_params_v2_vec[i].begin(),
+                    job_params_v2_vec[i].end());
+
+      result_mpi(ij).grad(z_var1, z_grad1);
+      result_concurrent(ij).grad(z_var2, z_grad2);
+
+      stan::math::set_zero_all_adjoints();
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/functor/map_rect_mpi_test.cpp]: This set-zero isn't doing anything at this point.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+ * valid output chunk since the gather commands issued on the root to
+ * collect results would otherwise fail. Thus, if a job fails on any
+ * worker the a respective status flag will be transferred after the
+ * reduce and the results gather step.
+ *
+ * Note 2: During the first evaluation of the function the ragged
+ * array sizes need to be collected on the root from all workers. This
+ * is needed on the root such that the root knows how many outputs are
+ * computed on each worker. This information is then cached for all
+ * subsequent evaluations. However, caching this information can only
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: oh, thanks for catching.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+    boost::mpi::broadcast(world_, local_data.data(), data_size[0], 0);
+
+    return local_data;
+  }
+
+  // scatters an Eigen matrix column wise over the cluster. Meta
+  // information as the data size is treated as static data and
+  // only transferred on the first call and read from cache
+  // subsequently.
+  template <int meta_cache_id>
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: Sure.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+ */
+template <int call_id, typename ReduceF, typename CombineF>
+class mpi_parallel_call {
+  boost::mpi::communicator world_;
+  const std::size_t rank_ = world_.rank();
+  const std::size_t world_size_ = world_.size();
+  std::unique_lock<std::mutex> cluster_lock_;
+
+  // local caches which hold local slices of data
+  typedef internal::mpi_parallel_call_cache<call_id, 1,
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I agree with the `_t` convention, but here I deliberately deviate from that since the type definition is the same as declaring data. This is odd, I know, but the typedef is a declaration of a data item which exists a single time in the program. That is why I would prefer to stay without the `_t` here.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+
+    job_chunk.reduce_combine();
+  }
+
+  // all data is cached and local parameters are also available
+  result_t reduce_combine() {
+    const std::vector<int>& job_chunks = cache_chunks::data();
+    const int num_jobs = sum(job_chunks);
+
+    // id of first job out of all
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: job_chunks has the size world_size which is the number of workers in the cluster and as such rank < job_chunks.size() is always fine. job_chunks contains the number of jobs assigned to a worker of some rank. Thus I need to sum all job_chunk entries up to (but excluding) rank as the current worker with the given rank has to process job_chunks[rank] jobs.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+    int start_job = 0;
+    for (std::size_t n = 0; n < rank_; ++n)
+      start_job += job_chunks[n];
+
+    const int num_local_jobs = local_job_params_dbl_.cols();
+    int local_outputs_per_job = num_local_jobs == 0 ? 0 : num_outputs_per_job_;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I think we want to allow that jobs *can* return no output. This is an odd function, but in Stan we allow for that. Thus we need `num_outputs_per_job` to be -1 to indicate that things are not yet initialized. Right?

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+    matrix_d local_output(
+        local_outputs_per_job == -1 ? 0 : local_outputs_per_job,
+        num_local_jobs);
+    std::vector<int> local_f_out(num_local_jobs, -1);
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: As I say, we need to initialize to a value such that I know that the function has not yet returned any useful result. So we need -1 which is not a meaningful output size ever.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+
+        if (local_output.cols() < offset + local_f_out[i])
+          local_output.conservativeResize(Eigen::NoChange,
+                                          2 * (offset + local_f_out[i]));
+
+        local_output.block(0, offset, local_output.rows(), local_f_out[i])
+            = job_output;
+      }
+    } catch (const std::exception& e) {
+      local_ok = 0;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I think I doced this in the doxygen comment above. See Note 1 there. Is that sufficient? I could add a comment which refers to note 1. The thing is that the exception must be thrown on the root only. Otherwise things get out of sync which leads to a catastrophe (undefined behavior).

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
+        }
+      }
+
+      // execution was ok everywhere such that we can now distribute the
+      // outputs per job as recorded on the root on the first execution
+      boost::mpi::broadcast(world_, local_outputs_per_job, 0);
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: Yes, there is. The root broadcasts it's value to everyone else. The workers overwrite the variable with the value from root. I cannot use the `num_outputs_per_job_` directly because it must be maintained at `-1` as long as all jobs from all workers were OK.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+      num_outputs_per_job_ = local_outputs_per_job;
+
+      if (!cache_f_out::is_valid()) {
+        // ok, now all is fine everywhere, collect sizes and cache info
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: that's fine. The code is subtle, I know that. I am glad you bear with me here.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+      }
+    }
+
+    typename cache_f_out::cache_t& world_f_out = cache_f_out::data();
+
+    // check that cached sizes are the same as just collected from
+    // this evaluation
+    for (int i = 0; i < num_local_jobs; ++i) {
+      if (world_f_out[start_job + i] != local_f_out[i]) {
+        local_ok = 0;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: You have to continue to keep everyone synced. Assume I would quit immediately. That would mean that one of the workers won't reach the gather statement below... and this would mean that the root would wait forever for the worker which did a premature quit. Therefore, the code is setup to keep running and only transmit the success status at the very end.

Things happen in different processes and communication needs to stay synched. That does not make things easier...

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp (hunk 1 above)
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: Here I am adding up the local_ok from all workers. If that sum is equal to the world_size, then (and only then) all workers had local_ok==1.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp (hunk 4 above)
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I can check again, but I think I need the `remove_cv` here. The complication here is that the same code does slightly different things when run on the root and on a worker. The root sends while the workers receive the data (using the same code chunk).

**Comment by wds15** [comment]
Thanks much for going through that massive code block. I will address them probably over the weekend.

**Comment by bob-carpenter** [comment]
Yes, that's fine.  It'd also be OK to put a poitner back to the note
in the catch block.  It just looked odd to me that there wasn't any
overt action taking on exception other than to set a flag.

**Comment by bob-carpenter** [comment]
> 
> I think we want to allow that jobs can return no output. This is an odd function, but in Stan we allow for that. Thus we need num_outputs_per_job to be -1 to indicate that things are not yet initialized. Right?

OK.

**Comment by bob-carpenter** [comment]
Although I don't understand the reasoning, if you feel that strongly about it, I won't insist---I don't feel that strongly about a _t suffix convention.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 16]
```diff
@@ -0,0 +1,488 @@
...
+      cache_f_out;
+  typedef internal::mpi_parallel_call_cache<call_id, 4, std::vector<int>>
+      cache_chunks;
+
+  // # of outputs for given call_id+ReduceF+CombineF case
+  static int num_outputs_per_job_;
+
+  CombineF combine_;
+
+  typedef typename CombineF::result_t result_t;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: OK

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 17]
```diff
@@ -0,0 +1,488 @@
...
+  template <typename T_shared_param, typename T_job_param>
+  mpi_parallel_call(
+      const Eigen::Matrix<T_shared_param, Eigen::Dynamic, 1>& shared_params,
+      const std::vector<Eigen::Matrix<T_job_param, Eigen::Dynamic, 1>>&
+          job_params,
+      const std::vector<std::vector<double>>& x_r,
+      const std::vector<std::vector<int>>& x_i)
+      : combine_(shared_params, job_params) {
+    if (rank_ != 0)
+      throw std::runtime_error("problem sizes can only defined on the root.");
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: OK

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+    check_matching_sizes("mpi_parallel_call", "job parameters", job_params,
+                         "continuous data", x_r);
+    check_matching_sizes("mpi_parallel_call", "job parameters", job_params,
+                         "integer data", x_i);
+
+    // in case we have already cached data available for this
+    // call_id, do further checks
+    if (cache_chunks::is_valid()) {
+      typename cache_chunks::cache_t& job_chunks = cache_chunks::data();
+      const int cached_num_jobs = sum(job_chunks);
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: OK

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+    cluster_lock_ = mpi_broadcast_command<stan::math::mpi_distributed_apply<
+        mpi_parallel_call<call_id, ReduceF, CombineF>>>();
+
+    const std::vector<int> job_dims = dims(job_params);
+
+    const size_type num_jobs = job_dims[0];
+    const size_type num_job_params = num_jobs == 0 ? 0 : job_dims[1];
+
+    const vector_d shared_params_dbl = value_of(shared_params);
+    matrix_d job_params_dbl(num_job_params, num_jobs);
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: but I am going from an array of Eigen vectors to a Eigen matrix. So the type changes and it's not merely a double conversion.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+ public:
+  // called on root
+  template <typename T_shared_param, typename T_job_param>
+  mpi_parallel_call(
+      const Eigen::Matrix<T_shared_param, Eigen::Dynamic, 1>& shared_params,
+      const std::vector<Eigen::Matrix<T_job_param, Eigen::Dynamic, 1>>&
+          job_params,
+      const std::vector<std::vector<double>>& x_r,
+      const std::vector<std::vector<int>>& x_i)
+      : combine_(shared_params, job_params) {
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: That one is anyway static... should we document that along with the constructor?

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+    if (rank_ != 0)
+      throw std::runtime_error("problem sizes can only defined on the root.");
+
+    check_matching_sizes("mpi_parallel_call", "job parameters", job_params,
+                         "continuous data", x_r);
+    check_matching_sizes("mpi_parallel_call", "job parameters", job_params,
+                         "integer data", x_i);
+
+    // in case we have already cached data available for this
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: yeah... sometimes is hard to judge for me what's trivial and what not. I dropped this one.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+    // call_id, do further checks
+    if (cache_chunks::is_valid()) {
+      typename cache_chunks::cache_t& job_chunks = cache_chunks::data();
+      const int cached_num_jobs = sum(job_chunks);
+      check_size_match("mpi_parallel_call", "cached number of jobs",
+                       cached_num_jobs, "number of jobs", job_params.size());
+    }
+
+    // make childs aware of upcoming job & obtain cluster lock
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: I can always say to be non-native speaker as an excuse (to some extent). Changed.

**Comment by bob-carpenter** [comment]
I think it may be able to deal with that.  I think you can assign Eigen matrices of different shapes to each other.  But it's no big deal either way---just a simple loop.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 23]
```diff
@@ -0,0 +1,488 @@
...
+    job_chunk.reduce_combine();
+  }
+
+  // all data is cached and local parameters are also available
+  result_t reduce_combine() {
+    const std::vector<int>& job_chunks = cache_chunks::data();
+    const int num_jobs = sum(job_chunks);
+
+    // id of first job out of all
+    int start_job = 0;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: OK

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 24]
```diff
@@ -0,0 +1,488 @@
...
+    for (std::size_t n = 0; n < rank_; ++n)
+      start_job += job_chunks[n];
+
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: sounds good.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+        num_local_jobs);
+    std::vector<int> local_f_out(num_local_jobs, -1);
+
+    typename cache_x_r::cache_t& local_x_r = cache_x_r::data();
+    typename cache_x_i::cache_t& local_x_i = cache_x_i::data();
+
+    // check if we know already output sizes
+    if (cache_f_out::is_valid()) {
+      typename cache_f_out::cache_t& f_out = cache_f_out::data();
+      int num_outputs = 0;
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: Indeed.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp [hunk 26]
```diff
@@ -0,0 +1,488 @@
...
+      bool all_ok = cluster_status == static_cast<int>(world_size_);
+      boost::mpi::broadcast(world_, all_ok, 0);
+      if (!all_ok) {
+        // err out on the root
+        if (rank_ == 0) {
+          throw std::domain_error("MPI error on first evaluation.");
+        } else {
+          // and ensure on the workers that they return into their
+          // listening state
+          return result_t();
```
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: The way I put it made it easier for me to read it, but if you think different... fine for me.

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp (hunk 2 above)
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: Uhmm... `dims` is a stan::math function which tells me the size of the data shapes (and not the first two entries). So I guess I should leave this as is; note that different things happen on the root (sending) and on the workers (receiving).

File: stan/math/prim/mat/functor/mpi_parallel_call.hpp (hunk 3 above)
Comment (wds15) [reply stan/math/prim/mat/functor/mpi_parallel_call.hpp]: OK

**Comment by wds15** [comment]
Lot's of more doc here now. In going over it in this iteration I found in a few places simplifications. I hope all is fine now (including the way the concepts are described for ReduceF and CombineF).

**Comment by bob-carpenter** [comment]
Thanks.  I should have time to get to finish the final review tomorrow (Thursday) afternoon or Friday.

**Comment by bob-carpenter** [comment]
@wds15 --- assuming the build is all OK for this to not torch Jenkins, go ahead and merge.  Everything looks great.
